engine/classes/Elgg/Http/ResponseFactory.php

```diff
@@ -0,0 +1,334 @@
+	 * @param array   $headers An array of response headers (Location is always set to the given URL)
+	 * @return RedirectResponse
+	 */
+	public function prepareRedirectResponse($url, $status = 302, array $headers = array()) {
+		$response = new RedirectResponse($url, $status, $headers);
+		$response->prepare($this->request);
+		return $response;
+	}
+
+	/**
+	 * Send HTTP response
+	 *
+	 * @param ResponseBuilder $response An instanceof an ErrorResponse, OkResponse or RedirectResponse
+	 * @return Response
+	 */
+	public function respond(ResponseBuilder $response) {
+
+		$response_type = $this->parseContext();
+		$response = $this->hooks->trigger('response', $response_type, $response, $response);
+
+		$is_xhr = $this->request->isXmlHttpRequest();
+
+		$is_action = false;
+		if (0 === strpos($response_type, 'action:')) {
+			$is_action = true;
+		}
+
+		if ($is_action && $response->getForwardURL() === null) {
+			// actions must always set a redirect url
+			$response->setForwardURL(REFERRER);
+		}
+
+		if ($response->getForwardURL() !== null && !$is_xhr && $response->isOK()) {
+			// non-xhr requests should issue a forward if redirect url is set
+			$response->setStatusCode(ELGG_HTTP_FOUND);
+		}
+
+		if ($is_xhr && $is_action && !$this->ajax->isAjax2Request()) {
+			// xhr actions using legacy ajax API should return 200 with wrapped data
+			$response->setStatusCode(ELGG_HTTP_OK);
+			$response->setContent($this->wrapLegacyAjaxResponse($content, $response->getForwardURL()));
+		}
+
+		$content = $this->stringify($response->getContent());
+		$status_code = $response->getStatusCode();
+		$headers = $response->getHeaders();
+
+		if ($response->isRedirection()) {
+			$redirect_url = $response->getForwardURL();
+			return $this->redirect($redirect_url, $status_code)->send();
+		}
+
+		if ($response->isClientError() || $response->isServerError()) {
+			return $this->respondWithError($content, $status_code, $headers);
+		}
+
+		return $this->respondFromContent($content, $status_code, $headers);
+	}
+
+	/**
+	 * Send error HTTP response
+	 *
+	 * @param string $error       Error message
+	 * @param int    $status_code HTTP status code
+	 * @param array  $headers     HTTP headers (will be discarded on AJAX requests)
+	 * @return Response
+	 */
+	public function respondWithError($error, $status_code = ELGG_HTTP_BAD_REQUEST, array $headers = []) {
+		if ($this->ajax->isReady()) {
+			return $this->ajax->respondWithError($error, $status_code);
+		}
+
+		$is_xhr = $this->request->isXmlHttpRequest();
+
+		if ($is_xhr) {
+			return $this->prepareResponse($error, $status_code, $headers)->send();
+		}
+
+		$is_action = false;
+		if (0 === strpos($response_type, 'action:')) {
+			$is_action = true;
+		}
+
+		$forward_url = $this->request->headers->get('Referer');
+
+		if (!$is_action) {
+			$params = [
+				'current_url' => current_page_url(),
+				'forward_url' => $forward_url,
+			];
+
+			$this->hooks->trigger('forward', (string) $status_code, $params, $url);
```

Comment:
Right. I think we should also trigger for actions with numeric status codes not in [400, 403, 404] to stay BC?